presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxBy.java

```diff
@@ -99,37 +143,81 @@ private InternalAggregationFunction generateAggregation(Type valueType, Type key
         return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INPUT_CHANNEL, key), new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void input(boolean min, Type keyType, MaxOrMinByState state, Block value, Block key, int position)
+    private void generateInputMethod(ClassDefinition definition, CallSiteBinder binder, MethodHandle compareMethod, Type keyType, Type valueType, Class<?> stateClass)
     {
-        if ((state.getKey() == null) || state.getKey().isNull(0) ||
-                compare(keyType.compareTo(key, position, state.getKey(), 0), min)) {
-            state.setKey(key.getSingleValueBlock(position));
-            state.setValue(value.getSingleValueBlock(position));
+        Parameter state = arg("state", stateClass);
+        Parameter value = arg("value", Block.class);
+        Parameter key = arg("key", Block.class);
+        Parameter position = arg("position", int.class);
+        MethodDefinition method = definition.declareMethod(a(PUBLIC, STATIC), "input", type(void.class), state, value, key, position);
+        if (keyType.equals(UNKNOWN)) {
+            method.getBody().ret();
+            return;
         }
-    }
-
-    public static void combine(boolean min, Type keyType, MaxOrMinByState state, MaxOrMinByState otherState)
-    {
-        Block key = state.getKey();
-        Block otherKey = otherState.getKey();
-        if ((key == null) || ((otherKey != null) && compare(keyType.compareTo(otherKey, 0, key, 0), min))) {
-            state.setKey(otherKey);
-            state.setValue(otherState.getValue());
+        BytecodeBlock body = method.getBody();
+        Class<?> keyJavaType = keyType.getJavaType();
+        SqlTypeBytecodeExpression keySqlType = constantType(binder, keyType);
+
+        BytecodeBlock ifBlock = new BytecodeBlock()
+                .append(state.invoke("setFirstState", void.class, keySqlType.getValue(key, position)))
+                .append(state.invoke("setFirstStateNull", void.class, constantBoolean(false)))
+                .append(state.invoke("setSecondStateNull", void.class, value.invoke("isNull", boolean.class, position)));
+        if (!valueType.equals(UNKNOWN)) {
+            SqlTypeBytecodeExpression valueSqlType = constantType(binder, valueType);
+            ifBlock.append(new IfStatement()
+                    .condition(value.invoke("isNull", boolean.class, position))
+                    .ifFalse(state.invoke("setSecondState", void.class, valueSqlType.getValue(value, position))));
         }
+
+        body.append(new IfStatement()
+                .condition(or(state.invoke("isFirstStateNull", boolean.class),
```

Comment:
Format function calls that span multiple lines as one argument per line:

```
                .condition(or(
                        state.invoke("isFirstStateNull", boolean.class),
                        and(
                                not(key.invoke("isNull", boolean.class, position)),
                                loadConstant(binder, compareMethod, MethodHandle.class).invoke("invokeExact", boolean.class, keySqlType.getValue(key, position), state.invoke("getFirstState", keyJavaType)))))
```

In this particular case, it makes the hierarchy of the logical expression immediately clear after reformatting.